TickSpec/FeatureParser.fs

```diff
@@ -1,92 +1,131 @@
+                | Some (stateTags, stateMap) ->
+                    let newStateMap =
+                        rowMap
+                        |> Map.fold (fun state key value ->
+                            match state with
+                            | None -> None
+                            | Some map ->
+                                let existingValue = map |> Map.tryFind key
+                                match existingValue with
+                                | None -> Some (map.Add (key, value))
+                                | Some x when x = value -> Some map
+                                | _ -> None) (Some stateMap)
+                    match newStateMap with
+                    | None -> None
+                    | Some s ->
+                        let newStateTags = stateTags @ rowTags
+                        Some (newStateTags, s)
+            ) (Some (List.Empty, Map.empty))
+
+        examples
+        |> Seq.map (fun exampleBlock ->
+            exampleBlock.Table.Header |> List.sort,
+            exampleBlock.Table.Rows
+            |> Seq.map (fun row ->
+                Seq.zip exampleBlock.Table.Header row
+                |> Seq.fold (fun map (header, value) ->
+                        match Map.tryFind header map with
+                        | Some x ->
+                            let m = sprintf "A single header was specified multiple times in an example block starting at row %d" exampleBlock.LineNumber
+                            ParseException(m, Some exampleBlock.LineNumber) |> raise
+                        | None -> map |> Map.add header value
+                    ) Map.empty,
+                exampleBlock.Tags
             )
         )
-    let bullets =
-        line.Bullets
-        |> Option.map (fun bullets -> bullets |> Array.map replace)
-    (scenario,n,tags,{line with Table=table;Bullets=bullets},step)
-
-/// Appends shared examples to scenarios as examples
-let internal appendSharedExamples (sharedExamples:Table[]) scenarios  =
-    if Seq.length sharedExamples = 0 then
-        scenarios
-    else
-        scenarios |> Seq.map (function
-            | scenarioName,tags,steps,None ->
-                scenarioName,tags,steps,Some(sharedExamples)
-            | scenarioName,tags,steps,Some(exampleTables) ->
-                scenarioName,tags,steps,Some(Array.append exampleTables sharedExamples)
+        // Union tables with the same columns
+        |> Seq.groupBy (fun (h,_) -> h)
+        |> Seq.map (fun (header,tables) ->
+            header, tables |> Seq.collect (fun (_, rows) -> rows) |> Seq.toList)
+        |> Seq.toList
+        // Cross-join tables with different columns
+        |> combinations
+        |> Seq.map processRow
+        |> Seq.choose id
+        |> Seq.groupBy (fun (_,r) -> r)
+        |> Seq.map (fun (row, taggedRows) ->
+            taggedRows
+            |> Seq.fold (fun tags (t,r) ->
+                Seq.append tags t
+            ) Seq.empty
```

Comment:
Consider: According to https://docs.microsoft.com/en-us/dotnet/fsharp/style-guide/formatting#formatting-function-parameter-application I think it is weird to continue argument on the same scope, I would prefer e.g.:
```FSharp
|> Seq.fold
    (fun tags (t,r) -> Seq.append tags t)
    Seq.empty 
```
you have above in new code much bigger Seq.fold uses with big lambda expression and a similar closing + adding last argument. The formatting guidelines tell "However, if the body of a lambda expression is more than one line, consider factoring it out into a separate function rather than have a multi-line construct applied as a single argument to a function."

Reply:
Again, the Seq.empty at the end is the problem for me (and I don't like the fun on next line

suggest 
```
(Seq.empty, taggedRows) ||> Seq.fold (fun (row, taggedRows) ->
    Seq.append tags t)
```

(If the line ended between 80-120 I'd consider keeping it on one line too)

That guide also mentions an important reason for the pipe-frward and/or double pipeforward is for the type inference to flow correctly

Reply:
Thanks, I have changed it a little. Does it look better?